### ISSUE TITLE
rtimer: Calling rtimer_set before previous rtimer has expired calls wrong function too early

### ISSUE THREAD

**Issue by gebart** [contiki-os/contiki#814]
Looking at the code for `rtimer_set` in core/sys/rtimer.c it seems like calling `rtimer_set` with a new timer before the first one has expired will cause the second callback function to be called at the time when the first timer was supposed to be called, and the first callback will simply be forgotten. Nothing will be called at the second timer's expiration, since `rtimer_arch_schedule` is never called again.

I believe this is a bug in the core rtimer library.

I am trying to use an rtimer to trigger etimer and clock updates on the Mulle platform (PR #776), but it makes the clock and etimers to stop working after using an rtimer for anything.

Below is the terminal output from the timer-test in examples/cc2538dk, running on my Mulle board:

    2014-10-10 11:52:57,609 - INFO # -----------------------------------------
    2014-10-10 11:52:57,612 - INFO # Rtimer Test, 1 sec (32768 rtimer ticks):
    2014-10-10 11:52:57,614 - INFO # =======================
    2014-10-10 11:52:57,618 - INFO # Now=75788 (clock = 148) - For=108556
    2014-10-10 11:52:57,624 - INFO # Task called at 76289 (clock = 148)

The clock callback is forgotten once the rtimer_set call is made from the test thread, and the rtimer callback is made too early, at the time of the next clock tick (76289) instead of at the rtimer target time (108556). After the above lines are printed the clock stops ticking, and the test app just hangs around waiting for an etimer that never will expire.

**Comment by yatch** [comment]
I also think this is a bug.... And it looks like #1290, a "nominated" PR, will resolve the issue by replacing the problematic code in rtimer.c.

**Comment by bkozak-scanimetrics** [comment]
> I also think this is a bug.... And it looks like #1290, a "nominated" PR, will resolve the issue by replacing the problematic code in rtimer.c.

As the author of #1290, I'm not sure that this should be considered a bug since the current rtimer implementation really doesn't support the use of multiple rtimers simultaneously at all and it doesn't seem that it is meant to either.

Even with #1290 the "buggy" behaviour is still present if ```RTIMER_MULTIPLE_ACCESS``` is false (and this is also the default). However with false ```RTIMER_MULTIPLE_ACCESS``` you would have the option to ```rtimer_cancel``` the old timer before setting a new one.

**Comment by simonduq** [comment]
I think we can basically close the issue, as @bkozak-scanimetrics mentioned, rtimer simply does not support multiple timers at the moment.

**Comment by yatch** [comment]
@bkozak-scanimetrics Oh, I overlooked the implementation of `add_single_access()` that is almost identical to the original `rtimer_set()`. My bad (>_<). You're right, rtimer behaves in the same way by default as it does now. I confirmed it with the tip of the branch for #1290,

I know the current rtimer does not support multiple timer tasks. I guess @gebart does as well. To my understanding, the point of this issue is the following part:

> calling rtimer_set with a new timer before the first one has expired will cause the second callback function to be called at the time when the first timer was supposed to be called

Simply speaking, the **second** timer callback function is called when the **first** timer expires. I think either of the following is an ideal result even if the rtimer supports a single timer task at most:

1. the **first** timer callback function is called when the **first** timer expires (the first wins)
2. the **second** timer callback function is called when the **second** timer expires (the last wins)

I prefer the former with making `rtimer_set()` for the second timer return an error code, for example, `RTIMER_ERR_FULL`.

Anyway, I'm OK with the closure of this old issue. If the above is the case, I'll open a separate issue to discuss it. What do you think...?

**Comment by simonduq** [comment]
OK, re-opening. You're right things can be improved even in the single-rtimer case. I like the first solution you suggest (second attempt fails), feel free to PR!

**Comment by yatch** [comment]
@simonduq Thank you for reopening it. I will send a PR with reference to this issue once #1290 is merged.
